From 325b7773c89b498de357d2952ed47ba052658296 Mon Sep 17 00:00:00 2001
From: Maximilian Weigand <mweigand@mweigand.net>
Date: Sat, 11 Jun 2022 20:58:17 +0200
Subject: [PATCH 20/21] [rockchip_ebc] Fix a bug in the scheduling function
 that could schedule an area too early: if the area overlaps with an
 already-started area, its begin_frame will be set to the end frame of the
 other one. However, if any frame in the list follows that can start earlier
 (because it does not overlap or finishes at an earlier time) than this
 earlier end frame will be used to schedule the new area.

---
 drivers/gpu/drm/rockchip/rockchip_ebc.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_ebc.c b/drivers/gpu/drm/rockchip/rockchip_ebc.c
index 7e1558403973..973d13ffd0d3 100644
--- a/drivers/gpu/drm/rockchip/rockchip_ebc.c
+++ b/drivers/gpu/drm/rockchip/rockchip_ebc.c
@@ -576,8 +576,9 @@ static bool rockchip_ebc_schedule_area(struct list_head *areas,
 			return false;
 		}
 
-		/* Otherwise, start at the same time as the other area. */
-		frame_begin = other->frame_begin;
+		/* Otherwise, the earliest start is the same time as that of the other
+		 * area. */
+		frame_begin = max(frame_begin, other->frame_begin);
 	}
 
 	area->frame_begin = frame_begin;
-- 
2.30.2

